File: modulesCore/homeScreen/src/main/java/cn/pivotstudio/modulec/homescreen/ui/adapter/AllForestItemAdapter.java
package cn.pivotstudio.modulec.homescreen.ui.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;
import cn.pivotstudio.modulec.homescreen.databinding.ItemForestCardBinding;
import cn.pivotstudio.modulec.homescreen.model.ForestCard;

public class AllForestItemAdapter extends ListAdapter<ForestCard, AllForestItemAdapter.ForestCardViewHolder> {

    private View.OnClickListener itemClickListener;

    public static final DiffUtil.ItemCallback<ForestCard> DIFF_CALLBACK = new DiffUtil.ItemCallback<ForestCard>() {
        @Override
        public boolean areItemsTheSame(@NonNull ForestCard oldItem, @NonNull ForestCard newItem) {
            return oldItem.getForestId() == newItem.getForestId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull ForestCard oldItem, @NonNull ForestCard newItem) {
            return oldItem.getName().equals(newItem.getName());
        }
    };

    public AllForestItemAdapter(View.OnClickListener clickListener) {
        super(DIFF_CALLBACK);
        itemClickListener = clickListener;
    }

    @NonNull
    @Override
    public ForestCardViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ForestCardViewHolder(
                ItemForestCardBinding.inflate(LayoutInflater.from(parent.getContext()),
                        parent,
                        false)
        );
    }

    @Override
    public void onBindViewHolder(@NonNull ForestCardViewHolder holder, int position) {
        ForestCard card = getItem(position);
        holder.itemView.setOnClickListener(itemClickListener);
        holder.bind(card);
    }

    static class ForestCardViewHolder extends RecyclerView.ViewHolder {
        private final ItemForestCardBinding binding;

        public ForestCardViewHolder(ItemForestCardBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(ForestCard forestCard) {
            binding.setForestCard(forestCard);
        }

    }

}

File: modulesCore/homeScreen/src/main/java/cn/pivotstudio/modulec/homescreen/ui/adapter/AllForestItemAdapter.kt
package cn.pivotstudio.modulec.homescreen.ui.adapter

import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import cn.pivotstudio.husthole.moduleb.network.model.ForestBrief
import cn.pivotstudio.modulec.homescreen.databinding.ItemForestBriefBinding

const val TAG = "AllForestItemAdapter"

class AllForestItemAdapter(
    val onItemClick: (String) -> Unit
) : ListAdapter<ForestBrief, AllForestItemAdapter.ForestBriefViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ForestBriefViewHolder {
        return ForestBriefViewHolder(
            ItemForestBriefBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            )
        )
    }

    override fun onBindViewHolder(holder: ForestBriefViewHolder, position: Int) {
        val card = getItem(position)
        holder.itemView.setOnClickListener {
            onItemClick(card.forestId)
        }
        holder.bind(card)
    }

    inner class ForestBriefViewHolder(private val binding: ItemForestBriefBinding) :
        RecyclerView.ViewHolder(
            binding.root
        ) {

        fun bind(brief: ForestBrief) {
            binding.forestBrief = brief
            binding.btnJoinForest.setOnClickListener {
                Log.d(TAG, "bind: Join Forest Btn tapped")
            }
        }
    }

    companion object {
        val DIFF_CALLBACK: DiffUtil.ItemCallback<ForestBrief> =
            object : DiffUtil.ItemCallback<ForestBrief>() {
                override fun areItemsTheSame(oldItem: ForestBrief, newItem: ForestBrief): Boolean {
                    return oldItem.forestId == newItem.forestId
                }

                override fun areContentsTheSame(oldItem: ForestBrief, newItem: ForestBrief): Boolean {
                    return oldItem == newItem
                }
            }
    }
}